flink-libraries/flink-sql-client/src/main/java/org/apache/flink/table/client/gateway/local/result/MaterializedCollectStreamResult.java

```diff
@@ -112,31 +155,64 @@ else if (!isRetrieving()) {
 	@Override
 	protected void processRecord(Tuple2<Boolean, Row> change) {
 		synchronized (resultLock) {
-			final Row row = change.f1;
 			// insert
 			if (change.f0) {
-				materializedTable.add(row);
-				rowPositionCache.put(row, materializedTable.size() - 1);
+				processInsert(change.f1);
 			}
 			// delete
 			else {
-				// delete the newest record first to minimize per-page changes
-				final Integer cachedPos = rowPositionCache.get(row);
-				final int startSearchPos;
-				if (cachedPos != null) {
-					startSearchPos = Math.min(cachedPos, materializedTable.size() - 1);
-				} else {
-					startSearchPos = materializedTable.size() - 1;
-				}
-
-				for (int i = startSearchPos; i >= 0; i--) {
-					if (materializedTable.get(i).equals(row)) {
-						materializedTable.remove(i);
-						rowPositionCache.remove(row);
-						break;
-					}
-				}
+				processDelete(change.f1);
 			}
 		}
 	}
+
+	@VisibleForTesting
+	protected List<Row> getMaterializedTable() {
+		return materializedTable;
+	}
+
+	// --------------------------------------------------------------------------------------------
+
+	private void processInsert(Row row) {
+		// limit the materialized table
+		if (materializedTable.size() - validRowPosition >= maxRowCount) {
+			cleanUp();
+		}
+		materializedTable.add(row);
+		rowPositionCache.put(row, materializedTable.size() - 1);
+	}
+
+	private void processDelete(Row row) {
+		// delete the newest record first to minimize per-page changes
+		final Integer cachedPos = rowPositionCache.get(row);
+		final int startSearchPos;
+		if (cachedPos != null) {
+			startSearchPos = Math.min(cachedPos, materializedTable.size() - 1);
+		} else {
+			startSearchPos = materializedTable.size() - 1;
+		}
+
+		for (int i = startSearchPos; i >= validRowPosition; i--) {
+			if (materializedTable.get(i).equals(row)) {
+				materializedTable.remove(i);
+				rowPositionCache.remove(row);
+				break;
+			}
+		}
+	}
+
+	private void cleanUp() {
+		// invalidate row
+		final Row deleteRow = materializedTable.get(0);
+		rowPositionCache.remove(deleteRow);
```

Comment:
Before clearing the cache, we might want to check if the cached index is equal to `validRowPosition`. Otherwise, we remove the cached index for duplicate of this row.

Reply:
this only works well if we update the cached indexes when removing the data.